### PR TITLE
fix: preserve typed Context in resource and prompt handlers

### LINKED ISSUE #3235: Parameterized Context loses request state in resource and prompt handlers
## Description

A resource-template or prompt handler annotated with a parameterized `Context[LifespanContextT]` receives a different `Context` object from the one the server injected. The reconstructed object has none of the private request state, so accessing `ctx.request_context`, `ctx.session`, `ctx.request_id`, or lifespan state fails with:

```text
ValueError: Context is not available outside of a request
```

The equivalent tool handler works, and handlers annotated with unparameterized `Context` happen to work. This makes the documented typed lifespan-context pattern unusable specifically in dynamic resources and prompts.

## Minimal reproduction

```python
from collections.abc import AsyncIterator
from contextlib import asynccontextmanager

import anyio
from mcp.client import Client
from mcp.server.mcpserver import Context, MCPServer


@asynccontextmanager
async def lifespan(_: MCPServer[str]) -> AsyncIterator[str]:
    yield "live-state"


async def main() -> None:
    server = MCPServer("probe", lifespan=lifespan)

    @server.resource("probe://{name}")
    async def resource(name: str, ctx: Context[str]) -> str:
        return f"{name}:{ctx.request_context.lifespan_context}"

    @server.prompt("probe")
    async def prompt(name: str, ctx: Context[str]) -> str:
        return f"{name}:{ctx.request_context.lifespan_context}"

    async with Client(server) as client:
        await client.read_resource("probe://value")
        await client.get_prompt("probe", {"name": "value"})


anyio.run(main)
```

Both calls fail. Each should return content containing `value:live-state`.

## Root cause

`ResourceTemplate.from_function` and `Prompt.from_function` correctly omit the detected context parameter from their public argument schemas, but then wrap the original handler with `pydantic.validate_call`. At invocation time they inject the live `Context` into the wrapped handler, so Pydantic validates it again.

For a generic annotation, Pydantic converts the unparameterized runtime instance into `Context[str, Any]`. That creates a new model instance and does not carry over `Context`'s private `_request_context`, `_mcp_server`, and related attributes:

```text
server-created Context
        |
        v
validate_call parameter: Context[str]
        |
        v
new Context[str, Any] instance
        |
        v
private request state absent
```

A direct identity probe on current `main` shows:

```text
Context       -> same object, request state present
Context[str]  -> new object, request state absent
```

Tools avoid this because `FuncMetadata.call_fn_with_arg_validation` validates only client-supplied arguments and passes injected values directly to the raw handler.

## Impact

This affects resource templates and prompts that use typed lifespan context, including otherwise valid code following the SDK's generic `Context[LifespanContextT]` typing. It can also hide in tests that only assert that `ctx` is non-null rather than reading request-scoped state.

A compatible fix should preserve the existing validation/coercion of client-supplied resource or prompt arguments while passing the server-created context object directly, with its identity and private state intact.

## Environment

- MCP Python SDK: 2.0.0 and current `main` at `a4f4ccd091138771535e17191123f20b30fda68e`
- Python: 3.12.13
- Pydantic: 2.12.5

AI assistance was used to investigate and draft this report.

### PR DESCRIPTION
**Pull request by morluto**
Fixes #3235.

### Problem

The server creates one live `Context` object for each request. That object carries private state linking it to the active request and session.

Resource-template and prompt handlers are wrapped with Pydantic's `validate_call`. When a handler uses the precise `Context[LifespanState]` annotation, Pydantic reconstructs the context as a new model instead of passing through the server-created object. Private request state is not part of that reconstructed model, so `ctx.request_context`, `ctx.session`, and related properties raise:

```text
ValueError: Context is not available outside of a request
```

```text
Before

Client arguments ─┐
Live Context ──────┴─► Pydantic validation ─► New Context ─► Handler error
                                                  │
                                                  └─ private request state is missing

After

Client arguments ─────► Pydantic validation ─┐
Live Context ─────────► pass through ─────────┴─► Handler
```

Bare `Context` happened to pass through unchanged, so the failure appeared only when resource or prompt handlers used `Context[LifespanState]`. Tool handlers were unaffected because they use a different argument-validation path.

### Change

The validation wrapper now marks only the SDK-injected context parameter with Pydantic's `SkipValidation`. Caller-supplied arguments still receive the existing validation and coercion, including parameters collected through `**kwargs`, while the original live context reaches synchronous and asynchronous handlers unchanged.

The wrapper restores the handler's public signature after Pydantic builds its validator, so introspection and generated schemas continue to expose the original types.

The lifespan and migration documentation now describes `Context[LifespanState]` as supported across tools, resource templates, and prompts; static resources do not support context injection.

### Regression coverage

The context-injection tests now use parameterized `Context` and access live request state through an asynchronous resource handler and a synchronous prompt handler. A resource-template regression also verifies that the exact context object is preserved while URI parameters continue to flow through `**kwargs`.

On the affected base revision, these tests fail when the reconstructed context first accesses its request state.

### Validation

- `uv run coverage run -m pytest -n auto` — 5,585 passed, 8 skipped, 1 xfailed
- `uv run coverage report` — 100.00% statement and branch coverage
- `uv run ruff format --check .` — passed
- `uv run ruff check .` — passed
- `uv run pyright` — passed
- `uvx --from pre-commit pre-commit run markdownlint --files docs/handlers/lifespan.md docs/migration.md` — passed

AI assistance was used to investigate and prepare this contribution; the final diff and validation results were reviewed before submission.

### CODE REVIEW

File: tests/server/mcpserver/test_server.py [hunk 1]
```diff
@@ -1197,9 +1197,9 @@ async def test_resource_with_context(self):
         mcp = MCPServer()
 
         @mcp.resource("resource://context/{name}")
-        def resource_with_context(name: str, ctx: Context) -> str:
+        async def resource_with_context(name: str, ctx: Context[None]) -> str:
```
[the remark this answers is not in the document]
Comment (morluto) [reply tests/server/mcpserver/test_server.py:1200]: Verified against the current v2 runtime: typing.get_type_hints returns the concrete class Context[NoneType, Any]; inspect.isclass is true, issubclass(annotation, Context) is true, and find_context_parameter returns "ctx". The focused typed-resource test and all 20 CI matrix cells pass, so no detection change is needed here.

File: docs/handlers/lifespan.md [hunk 2]
```diff
@@ -41,7 +41,7 @@ Nothing new. `ctx` is a **Context** parameter, so the SDK injects it and it neve
 
 `genre` is the only argument the model can pass. The lifespan is your server's business.
 
-`@mcp.resource()` and `@mcp.prompt()` functions can take a `ctx` parameter too, written as a bare `Context` for a reason the next section gets to. Everything `ctx` carries is in **[The Context](context.md)**.
+`@mcp.resource()` and `@mcp.prompt()` functions can take the same typed `ctx` parameter. Everything `ctx` carries is in **[The Context](context.md)**.
```
[the remark this answers is not in the document]
Comment (morluto) [reply docs/handlers/lifespan.md]: Good catch. Commit 543c7335 now limits this guidance to URI-template resource handlers, explicitly states that static resources cannot accept Context, and makes the recap use the same scope. Markdown lint and the docs check pass.

File: docs/handlers/lifespan.md [hunk 3]
```diff
@@ -41,7 +41,7 @@ Nothing new. `ctx` is a **Context** parameter, so the SDK injects it and it neve
 
 `genre` is the only argument the model can pass. The lifespan is your server's business.
 
-`@mcp.resource()` and `@mcp.prompt()` functions can take a `ctx` parameter too, written as a bare `Context` for a reason the next section gets to. Everything `ctx` carries is in **[The Context](context.md)**.
+URI-template `@mcp.resource("scheme://{param}")` handlers and `@mcp.prompt()` handlers can take the same typed `ctx` parameter. Static resources cannot accept `Context` because context injection is available only for resource templates. Everything `ctx` carries is in **[The Context](context.md)**.
```
[the remark this answers is not in the document]
Comment (morluto) [reply docs/handlers/lifespan.md:44]: Fixed in 34816ecc. The Context guide now scopes injection to URI-template resource handlers and explicitly states that static resource handlers cannot accept Context, both in the introduction and recap.

File: tests/server/mcpserver/resources/test_resource_template.py [hunk 4]
```diff
@@ -221,6 +221,39 @@ def my_func(key: str, value: int) -> dict[str, Any]:
         data = json.loads(content)
         assert data == {"key": "foo", "value": 123}
 
+    @pytest.mark.anyio
+    async def test_create_resource_preserves_context_with_variadic_kwargs(self):
+        """Injected context bypasses validation without changing **kwargs handling."""
+        context = Context()
+
+        def my_func(ctx, **kwargs: str):  # type: ignore
```
[the remark this answers is not in the document]
Comment (morluto) [reply tests/server/mcpserver/resources/test_resource_template.py]: Fixed in 34816ecc. The regression now annotates the handler as `ctx: Context[None]` while passing the unparameterized `Context()` object created by the server. That reproduces the original Pydantic reconstruction: the identity check fails on `origin/main` and passes on this branch. The type suppressions are also gone; the focused context suite, Pyright, and Ruff pass.

File: tests/server/mcpserver/test_server.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (morluto) [reply tests/server/mcpserver/test_server.py:1200]: This is already handled by the existing direct-subclass check. In this SDK, `Context` is a Pydantic generic model, so `typing.get_type_hints()` materializes `Context[None]` as the concrete class `Context[NoneType, Any]`: `inspect.isclass(annotation)` and `issubclass(annotation, Context)` are both true, while `typing.get_origin(annotation)` is `None`. A direct probe returns `find_context_parameter(handler) == "ctx"`, and the end-to-end resource/prompt tests pass. Adding an origin check would not affect this case.
